Add lint to suggest `into()` to construct reference-counted slice

rudyardrichter opened the pull request:
To resolve #2623.

rudyardrichter commented:
Hi @oli-obk, here's what I started on for #2623. A few questions—how can I print debug information from running the lint? I tried something like this:
```rust
cx.sess().span_note(expr.span, &format!("{:?}", ty))
```
(from that article linked in `CONTRIBUTING.md`) but it looks like that's outdated. I'm not positive though since it seems like this doesn't actually run the lint at all when I do either of these:
```
TESTNAME=ui/use_shared_from_slice cargo test --test compile-test

cargo run --bin clippy-driver -- -L ./target/debug tests/ui/use_shared_from_slice.rs
```
What else do I need to change to get the lint running?

oli-obk commented:
The dogfood test complains about

```
warning: unused import: `LintContext`
 --> clippy_lints/src/use_shared_from_slice.rs:2:57
  |
2 | use rustc::lint::{LateContext, LateLintPass, LintArray, LintContext, LintPass};
  |                                                         ^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
error: useless use of `format!`
  --> clippy_lints/src/use_shared_from_slice.rs:59:21
   |
59 |                     format!("TODO"),
   |                     ^^^^^^^^^^^^^^^ help: consider using .to_string(): `"TODO".to_string()`
   |
   = note: `-D useless-format` implied by `-D clippy`
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.205/index.html#useless_format
```

rudyardrichter commented:
Sorry, I had committed it after removing some things for debugging. After 5958ddf the tests are passing but I don't think it runs the lint (I don't see anything happen even if I put a `panic` in `check_expr`).

oli-obk commented:
You need to add the lint to lib.rs.

Also try running the utils/update_lints.py script first

rudyardrichter commented:
Yeah I just ran `util/update_lints.py` and didn't actually change `lib.rs` myself. It looks correct [here](https://github.com/rust-lang-nursery/rust-clippy/pull/2807/files#diff-99f9e44d09f3d77a654506c5efaefc41R914)

rudyardrichter commented:
@oli-obk would you mind checking up on the progress here? I fixed up `check_expr`  and moved the `match_type_parameter` out from a different lint to the utils to use it for this one, and added the `check_local` for checking type declarations.

I noticed `LateLintPass` has a `check_struct_field` function and thought I might add the same type declaration check for struct fields, but I couldn't find much relevant code or documentation to figure that out, so I've left that out.

oli-obk commented on clippy_lints/src/use_shared_from_slice.rs:
```diff
@@ -0,0 +1,135 @@
+        if let Some(arg) = args.get(0);
+
+        then {
+            if arg_is_vec {
```
So this is a little dangerous, because the message might appear even if the user has no way to change the type. For `Rc<Vec<[T]>>` this is is kinda fine, because the suggestion will still work, but for `Rc<String>` I'm not sure if `.as_str().into()` is equivalent.

phansch commented:
ping @rudyardrichter, do you still want to finish this up?

rudyardrichter commented:
@phansch Yeah, I will get back to this and try running this lint like @oli-obk suggested. I may need some suggestions for how to lint the `Rc<String>` case

flip1995 commented:
Ping @rudyardrichter. I'm going over old PRs, that were abandoned by us reviewers (sorry for that!) or by the authors. Are you still interested in completing this?

rudyardrichter commented:
Hey @flip1995, probably best to let someone else take this. Apologies.

flip1995 commented:
No worries. Sorry for not giving you an earlier review. Thanks for your contribution!

HKalbasi commented:
finished in #6044
@rustbot label -S-inactive-closed
